src/asn1.ml

```diff
@@ -615,11 +667,263 @@ struct
+    in
+    let tokens = tokenize ~acc:[] bytes 0 (Cstruct.len bytes) None Init in
+    let rec parse = function
+      | `Type (_, (`PUBLIC_KEY, _)) :: `Length _ :: `Value ls :: _ ->
+          parse ls
+      | `Type (_, (`OID, _)) :: `Length _ :: `Bytes bytes :: tl ->
+          let bytes =
+            let prefix =
+              Printf.sprintf "\006%c" (Char.chr (Cstruct.len bytes))
+              |> Cstruct.of_string
+            in
+            Cstruct.append prefix bytes
+          in
+          `Oid (decode_oid bytes) :: parse tl
+      | `Type (_, (`MODULUS, _)) :: `Length _ :: `Bytes bytes :: tl ->
+          `Modulus (atoz_bigendian bytes) :: parse tl
+      | `Type (0x82, ((*`EXPONENT*) _ , _)) :: `Length _ :: `Bytes bytes :: tl ->
+          `Exponent (atoz_bigendian bytes) :: parse tl
+      | `Type (_, (`COEFFICIENT_B, _)) :: `Length _ :: `Bytes bytes :: tl ->
+          `Coefficient_b (atoz_bigendian bytes) :: parse tl
+      | `Type (_, (`BASE_POINT_G, _)) :: `Length _ :: `Bytes bytes :: tl ->
+          `Base_point_g (atoz_bigendian bytes) :: parse tl
+      | `Type (_, (`BASE_POINT_R_ORDER, _)) :: `Length _ :: `Bytes bytes :: tl ->
+          `Base_point_r_order (atoz_bigendian bytes) :: parse tl
+      | `Type (_, (`PUBLIC_POINT_Y, _)) :: `Length _ :: `Bytes bytes :: tl ->
+          `Public_point_y (atoz_bigendian bytes) :: parse tl
+      | `Type (_, (`COFACTOR_F, _)) :: `Length _ :: `Bytes bytes :: tl ->
+          `Cofactor_f (atoz_bigendian bytes) :: parse tl
+      | [] ->
+          []
+      | `Type (t, _) :: tl ->
+          let () = Printf.printf "`Type (0x%x, _) :: tl found\n" t in
+          parse tl
+      | `Length _ :: tl ->
+          let () = Printf.printf "`Length _ :: tl found\n" in
+          parse tl
+      | `Bytes _ :: tl ->
+          let () = Printf.printf "`Bytes _ :: tl found\n" in
+          parse tl
+      | `Value _ :: tl ->
+          let () = Printf.printf "`Value _ :: tl found\n" in
+          parse tl
+      | _ ->
+          raise (Failure "unaccounted for case")
+    in
+    let symbols = parse tokens in
+    let oid =
+      try
+        List.find (function `Oid x -> true | _ -> false) symbols
+      with Not_found -> `Oid (Unknown (Asn.OID.of_string "0.0.0.0.0.0.0.0.0.0"))
+    in
+    let open Result in
+    match oid with
+      | `Oid RSA ->
+          Ok (
+            `RSA
+              RSA_CVC.Public.(
+                let n =
+                  List.fold_left (fun acc x -> (match x with `Modulus x -> [x] | _ -> []) @ acc) [] symbols
+                  |> List.hd
+                in
+                let e =
+                  List.fold_left (fun acc x -> (match x with `Exponent x -> [x] | _ -> []) @ acc) [] symbols
+                  |> List.hd
+                in
+                {n; e}))
+      | `Oid ECDSA ->
+          Ok
+            (`ECDSA
+              ECDSA_CVC.Public.(
+                let modulus =
```

Comment:
These entries, in a valid key, are always all present sorted by tag number. So you can do something like:

```ocaml
match symbols with
| [ `Modulus modulus
  ; `Coefficient_a coefficient_a
  ; ...
  ] -> Result.Ok { modulus ; coefficient a ; ... }
| _ -> Result.Error "Parse error: some elements are missing or are not correctly sorted"
```
```